### PR TITLE
Batch verification for Groth16 and preparation for batch verification gadgets

### PR DESCRIPTION
**Pull request by weikengchen**
## Motivation

This PR adds the Groth16's batch verification algorithm as Pratyush mentioned in the `eng-snarkos` [sic] channel on June 29th, as well as some housekeeping changes planned for the future.

Since the writing platform is different (arkworks-rs vs zexe), some tuning has been done. The code tries to use the projective coordinates as much as possible, for efficiency.

In the future, we may run batch verification in the constraint worlds, which I hit some blockers and would need more discussion (e.g., generating the randomness), and it will affect the current Marlin implementation, which will conflict with `testnet2-dpc`, so I will continue with them later.

## Test Plan

It passes all the tests in snarkVM.

### CODE REVIEW

**Comment by weikengchen** [comment]
Converted to draft as there appear to be conflicts.

**Comment by weikengchen** [comment]
Ready for review. 

To Howard: use squash-and-merge, which will make the git history look nicer!

File: algorithms/src/snark/groth16/verifier.rs [hunk 1]
```diff
@@ -61,3 +65,82 @@ pub fn verify_proof<E: PairingEngine>(
...
+        .zip(&r_powers)
+        .map(|(proof, r)| proof.a.into_projective().mul(*r))
+        .collect::<Vec<_>>();
+    let combined_proof_a_s = E::G1Projective::batch_normalization_into_affine(combined_proof_a_s);
+    let ml_inputs = proofs
+        .iter()
+        .zip(&combined_proof_a_s)
+        .map(|(proof, a)| ((*a).prepare(), proof.b.prepare()))
+        .collect::<Vec<_>>();
+    let ml_inputs_ref = ml_inputs.iter().map(|(a, b)| (a, b)).collect::<Vec<_>>();
```
Comment (ljedrz) [inline algorithms/src/snark/groth16/verifier.rs:114]:
```suggestion
```

File: algorithms/src/snark/groth16/verifier.rs [hunk 2]
```diff
@@ -61,3 +65,82 @@ pub fn verify_proof<E: PairingEngine>(
...
+    let a_r_times_b = E::miller_loop(ml_inputs_ref.iter().cloned());
```
Comment (ljedrz) [inline algorithms/src/snark/groth16/verifier.rs:115]:
```suggestion
    let a_r_times_b = E::miller_loop(ml_inputs.iter().map(|(a, b)| (a, b)));
```

File: algorithms/src/snark/groth16/verifier.rs [hunk 3]
```diff
@@ -61,3 +65,82 @@ pub fn verify_proof<E: PairingEngine>(
...
+
+    let combined_c_s = {
+        let randomized_cs = proofs
```
Comment (ljedrz) [inline algorithms/src/snark/groth16/verifier.rs:118]:
```suggestion
        let mut randomized_cs = proofs
```

File: algorithms/src/snark/groth16/verifier.rs [hunk 4]
```diff
@@ -61,3 +65,82 @@ pub fn verify_proof<E: PairingEngine>(
...
+            .iter()
+            .zip(&r_powers)
+            .map(|(proof, r)| proof.c.into_projective().mul(*r))
+            .collect::<Vec<E::G1Projective>>();
```
Comment (ljedrz) [inline algorithms/src/snark/groth16/verifier.rs:122]:
```suggestion
```

File: algorithms/src/snark/groth16/verifier.rs [hunk 5]
```diff
@@ -61,3 +65,82 @@ pub fn verify_proof<E: PairingEngine>(
...
+        .map(|(proof, a)| ((*a).prepare(), proof.b.prepare()))
+        .collect::<Vec<_>>();
+    let ml_inputs_ref = ml_inputs.iter().map(|(a, b)| (a, b)).collect::<Vec<_>>();
+    let a_r_times_b = E::miller_loop(ml_inputs_ref.iter().cloned());
+
+    let combined_c_s = {
+        let randomized_cs = proofs
+            .iter()
+            .zip(&r_powers)
+            .map(|(proof, r)| proof.c.into_projective().mul(*r))
```
Comment (ljedrz) [inline algorithms/src/snark/groth16/verifier.rs:121]:
```suggestion
            .map(|(proof, r)| proof.c.into_projective().mul(*r));
```

File: algorithms/src/snark/groth16/verifier.rs [hunk 6]
```diff
@@ -61,3 +65,82 @@ pub fn verify_proof<E: PairingEngine>(
...
+            .collect::<Vec<E::G1Projective>>();
+
+        let mut combined_c_s = randomized_cs[0];
+        for randomized_c in randomized_cs.iter().skip(1) {
+            combined_c_s += randomized_c;
+        }
+        combined_c_s
```
Comment (ljedrz) [inline algorithms/src/snark/groth16/verifier.rs:128]:
```suggestion
        let mut randomized_cs = proofs
            .iter()
            .zip(&r_powers)
            .map(|(proof, r)| proof.c.into_projective().mul(*r));

        let first_rand = randomized_cs.next().unwrap();

        randomized_cs.fold(first_rand, |acc, rand| acc + rand)
```

File: algorithms/src/snark/groth16/verifier.rs [hunk 7]
```diff
@@ -61,3 +65,82 @@ pub fn verify_proof<E: PairingEngine>(
...
+                }
+                g_ic.mul(*r)
+            })
+            .collect::<Vec<E::G1Projective>>();
+
+        let mut combined_inputs = randomized_inputs[0];
+        for randomized_input in randomized_inputs.iter().skip(1) {
+            combined_inputs += randomized_input;
+        }
+        combined_inputs
```
Comment (ljedrz) [inline algorithms/src/snark/groth16/verifier.rs:100]:
```suggestion
        let mut randomized_inputs = public_inputs
            .iter()
            .zip(&r_powers)
            .map(|(input, r)| {
                let mut g_ic = pvk.vk.gamma_abc_g1[0].into_projective();
                for (i, b) in input.iter().zip(pvk.vk.gamma_abc_g1.iter().skip(1)) {
                    g_ic += &b.into_projective().mul(*i);
                }
                g_ic.mul(*r)
            });

        let first_rand = randomized_inputs.next().unwrap();

        randomized_inputs.fold(first_rand, |acc, rand| acc + rand)
```

File: gadgets/src/algorithms/snark/gm17.rs [hunk 8]
```diff
@@ -92,6 +96,93 @@ pub struct GM17PreparedVerifyingKeyGadget<Pairing: PairingEngine, F: Field, P: P
...
+
+            let g_alpha = P::G1Gadget::alloc(&mut cs.ns(|| "g_alpha"), || Ok(pvk.g_alpha.into_projective()))?;
+            let h_beta = P::G2Gadget::alloc(&mut cs.ns(|| "h_beta"), || Ok(pvk.h_beta.into_projective()))?;
+            let g_alpha_pc = P::G1PreparedGadget::alloc(&mut cs.ns(|| "g_alpha_pc"), || Ok(pvk.g_alpha.prepare()))?;
+            let h_beta_pc = P::G2PreparedGadget::alloc(&mut cs.ns(|| "h_beta_pc"), || Ok(pvk.h_beta.prepare()))?;
+            let g_gamma_pc = P::G1PreparedGadget::alloc(&mut cs.ns(|| "g_gamma_pc"), || Ok(pvk.g_gamma_pc.clone()))?;
+            let h_gamma_pc = P::G2PreparedGadget::alloc(&mut cs.ns(|| "h_gamma_pc"), || Ok(pvk.h_gamma_pc.clone()))?;
+            let h_pc = P::G2PreparedGadget::alloc(&mut cs.ns(|| "h_pc"), || Ok(pvk.h_pc.clone()))?;
+
+            let mut query = Vec::new();
```
Comment (ljedrz) [inline gadgets/src/algorithms/snark/gm17.rs:122]:
```suggestion
            let mut query = Vec::with_capacity(pvk.vk.query.len());
```

File: gadgets/src/algorithms/snark/gm17.rs [hunk 9]
```diff
@@ -92,6 +96,93 @@ pub struct GM17PreparedVerifyingKeyGadget<Pairing: PairingEngine, F: Field, P: P
...
+            let g_alpha_pc =
+                P::G1PreparedGadget::alloc_input(&mut cs.ns(|| "g_alpha_pc"), || Ok(pvk.g_alpha.prepare()))?;
+            let h_beta_pc = P::G2PreparedGadget::alloc_input(&mut cs.ns(|| "h_beta_pc"), || Ok(pvk.h_beta.prepare()))?;
+            let g_gamma_pc =
+                P::G1PreparedGadget::alloc_input(&mut cs.ns(|| "g_gamma_pc"), || Ok(pvk.g_gamma_pc.clone()))?;
+            let h_gamma_pc =
+                P::G2PreparedGadget::alloc_input(&mut cs.ns(|| "h_gamma_pc"), || Ok(pvk.h_gamma_pc.clone()))?;
+            let h_pc = P::G2PreparedGadget::alloc_input(&mut cs.ns(|| "h_pc"), || Ok(pvk.h_pc.clone()))?;
+
+            let mut query = Vec::new();
```
Comment (ljedrz) [inline gadgets/src/algorithms/snark/gm17.rs:165]:
```suggestion
            let mut query = Vec::with_capacity(pvk.vk.query.len());
```

File: gadgets/src/algorithms/snark/groth16.rs [hunk 10]
```diff
@@ -99,6 +101,84 @@ pub struct PreparedVerifyingKeyGadget<
...
+            let pvk = pvk.borrow().clone();
+
+            let alpha_g1_beta_g2 =
+                P::GTGadget::alloc(&mut cs.ns(|| "alpha_g1_beta_g2"), || Ok(pvk.alpha_g1_beta_g2.clone()))?;
+            let gamma_g2_neg_pc =
+                P::G2PreparedGadget::alloc(&mut cs.ns(|| "gamma_g2_neg_pc"), || Ok(pvk.gamma_g2_neg_pc.clone()))?;
+            let delta_g2_neg_pc =
+                P::G2PreparedGadget::alloc(&mut cs.ns(|| "delta_g2_neg_pc"), || Ok(pvk.delta_g2_neg_pc.clone()))?;
+
+            let mut gamma_abc_g1 = Vec::new();
```
Comment (ljedrz) [inline gadgets/src/algorithms/snark/groth16.rs:129]:
```suggestion
            let mut gamma_abc_g1 = Vec::with_capacity(pvk.vk.gamma_abc_g1.len());
```

File: gadgets/src/algorithms/snark/groth16.rs [hunk 11]
```diff
@@ -99,6 +101,84 @@ pub struct PreparedVerifyingKeyGadget<
...
+            let pvk = pvk.borrow().clone();
+
+            let alpha_g1_beta_g2 =
+                P::GTGadget::alloc_input(&mut cs.ns(|| "alpha_g1_beta_g2"), || Ok(pvk.alpha_g1_beta_g2.clone()))?;
+            let gamma_g2_neg_pc =
+                P::G2PreparedGadget::alloc_input(&mut cs.ns(|| "gamma_g2_neg_pc"), || Ok(pvk.gamma_g2_neg_pc.clone()))?;
+            let delta_g2_neg_pc =
+                P::G2PreparedGadget::alloc_input(&mut cs.ns(|| "delta_g2_neg_pc"), || Ok(pvk.delta_g2_neg_pc.clone()))?;
+
+            let mut gamma_abc_g1 = Vec::new();
```
Comment (ljedrz) [inline gadgets/src/algorithms/snark/groth16.rs:164]:
```suggestion
            let mut gamma_abc_g1 = Vec::with_capacity(pvk.vk.gamma_abc_g1.len());
```

File: gadgets/src/traits/algorithms/snark.rs [hunk 12]
```diff
@@ -25,6 +25,34 @@ use crate::{
...
+            RandProvider::PowerSeries(challenge) => {
+                let mut cur = F::one();
+                res.push(cur);
+
+                for _ in 1..num {
+                    cur *= challenge;
+                    res.push(cur)
+                }
+            }
+        }
```
Comment (ljedrz) [inline gadgets/src/traits/algorithms/snark.rs:46]:
```suggestion
        let RandProvider::PowerSeries(challenge) = self;

        let mut cur = F::one();
        res.push(cur);

        for _ in 1..num {
            cur *= challenge;
            res.push(cur)
        }
```

**Review by ljedrz — COMMENTED** [review]
Left a few suggestions that reduce the number of allocations and make the code more compact.

File: algorithms/src/snark/groth16/verifier.rs [hunk 13]
```diff
@@ -42,15 +46,15 @@ pub fn verify_proof<E: PairingEngine>(
         return Err(SynthesisError::MalformedVerifyingKey);
     }
 
-    let mut g_ic = pvk.gamma_abc_g1()[0];
+    let mut g_ic = pvk.gamma_abc_g1()[0].into_projective();
     for (i, b) in public_inputs.iter().zip(pvk.gamma_abc_g1().iter().skip(1)) {
-        g_ic.add_assign(b.mul(*i));
+        g_ic.add_assign(b.into_projective().mul(*i));
```
Comment (Pratyush) [inline algorithms/src/snark/groth16/verifier.rs:51]: This is going to be slower. Is there an issue with the original?

File: algorithms/src/snark/groth16/verifier.rs [hunk 14]
```diff
@@ -61,3 +65,82 @@ pub fn verify_proof<E: PairingEngine>(
...
+    }
+
+    let combined_inputs = {
+        let randomized_inputs = public_inputs
+            .iter()
+            .zip(&r_powers)
+            .map(|(input, r)| {
+                let mut g_ic = pvk.vk.gamma_abc_g1[0].into_projective();
+                for (i, b) in input.iter().zip(pvk.vk.gamma_abc_g1.iter().skip(1)) {
+                    g_ic += &b.into_projective().mul(*i);
```
Comment (Pratyush) [inline algorithms/src/snark/groth16/verifier.rs:90]:
```suggestion
                    g_ic += &b.mul(*i);
```

**Comment by howardwu** [comment]
@Pratyush thanks for calling those out. It was my bad introducing those. We should definitely revert the affine-based calls.

**Comment by weikengchen** [comment]
Nope, it is hard to decide, since in the native world, projective coordinates are often faster.

Since every curve has affine and projective, maybe it would be convenient if we are working on an interface called "Curve", which provides both interfaces.

**Comment by howardwu** [comment]
@weikengchen agreed on the Curve trait. The issue Pratyush is calling out is that our repr is in affine form and pay the cost of going to projective and back (going back is 2 inversions) every time we perform an operation.

**Pull request closed without merging**